tracing: fix zipkin annotation timestamp serialization issue

douglas-reid opened the pull request:
Commit Message: fix handling of timestamps in zipkin annotations / logs
Additional Description: the intended replacement from string to int was not happening as expected. this causes ingestion issues with non-lenient collectors, such as jaeger.  Augmented existing tests to use different timestamps for annotations to catch issue and prevent regression.
Risk Level: Low
Testing: Unit
Docs Changes: None
Release Notes: None
Platform Specific Features: None

Fixes #15960 

Signed-off-by: Douglas Reid <douglas-reid@users.noreply.github.com>

jamesmulcahy commented:
As we updated to v1.18.2 we ran into this issue too. One of our integration tests validates the zipkin integration, but for various reasons, that hadn't been run in conjunction with the original patch.  The test validates the data against the zipkin-go objects, and was running into the same string/int issue.

We've tested this fix this morning and confirmed that the integration test now passes. 

Thanks @douglas-reid.  From a review perspective, you may want to add an entry to the bugfixes section of the release notes?

jamesmulcahy commented:
I also wonder if we should get this into a v1.18.3 -- others may well run into this too.

douglas-reid commented:
> As we updated to v1.18.0 we ran into this issue too. One of our integration tests validates the zipkin integration, but for various reasons, that hadn't been run in conjunction with the original patch. The test validates the data against the zipkin-go objects, and was running into the same string/int issue.
> 
> We've tested this fix this morning and confirmed that the integration test now passes.
> 
> Thanks @douglas-reid. From a review perspective, you may want to add an entry to the bugfixes section of the release notes?
> 
> Hat Tip to @drobertduke from my team, who actually did the work/testing that I mentioned above -- I'm just the messenger.

Added to the release notes.

Agree that we should get this backported as well.

douglas-reid commented:
@dio any chance for a quick review on this PR?  It is a release-blocker for other projects.

douglas-reid commented:
/retest

htuch commented on docs/root/version_history/current.rst:
```diff
@@ -13,6 +13,8 @@ Bug Fixes
 ---------
 *Changes expected to improve the state of the world and are unlikely to have negative effects*
 
+* zipkin: fix timestamp serializaiton in annotations. A prior bug fix exposed an issue with timestamps being serialized as strings.
```
Nit: typo on serialization

douglas-reid replied:
fixed.

htuch commented on source/extensions/tracers/zipkin/span_buffer.cc:
```diff
@@ -133,7 +133,7 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span, Util::Replacements& rep
       (*annotation_entry_fields)[ANNOTATION_VALUE] = ValueUtil::stringValue(annotation.value());
       (*annotation_entry_fields)[ANNOTATION_TIMESTAMP] =
-          Util::uint64Value(annotation.timestamp(), annotation.value(), replacements);
+          Util::uint64Value(annotation.timestamp(), ANNOTATION_TIMESTAMP, replacements);
```
@dio does this make sense?

dio replied:
I think this is right. Sorry for the delay.

htuch reviewed (commented):
LGTM but I'd like @dio to verify behavior intent.

dio reviewed (dismissed):
Thanks, @douglas-reid sorry for the delay. LGTM. If you fix the DCO, I'll re-approve this.

htuch commented:
@douglas-reid please fix DCO and we can merge!

douglas-reid commented:
> @douglas-reid please fix DCO and we can merge!

well, trying to fix the DCO has certainly cause quite a mess. please hold while I try to figure out what is happening.

douglas-reid commented:
Trying again.

douglas-reid commented:
@htuch @dio had to move to a new PR because the DCO cluster. please accept my apologies. please review / approve https://github.com/envoyproxy/envoy/pull/16072.

douglas-reid commented:
Sorry, https://github.com/envoyproxy/envoy/pull/16073.  Today is not my day typing-wise.
